From: Jan Kara <jack@suse.cz>
Subject: [PATCH] writeback: Do not sort b_io list only because of block device
 inode
References: bnc#819018
Patch-mainline: submitted for 3.11

It is very likely that block device inode will be part of BDI dirty list
as well. However it doesn't make sence to sort inodes on the b_io list
just because of this inode (as it contains buffers all over the device
anyway). So save some CPU cycles which is valuable since we hold relatively
contented wb->list_lock.

Signed-off-by: Jan Kara <jack@suse.cz>

Index: linux-3.0-SLE11-SP3/fs/fs-writeback.c
===================================================================
--- linux-3.0-SLE11-SP3.orig/fs/fs-writeback.c
+++ linux-3.0-SLE11-SP3/fs/fs-writeback.c
@@ -245,6 +245,8 @@ static bool inode_dirtied_after(struct i
 	return ret;
 }
 
+extern struct super_block *blockdev_superblock;
+
 /*
  * Move expired dirty inodes from @delaying_queue to @dispatch_queue.
  */
@@ -263,10 +265,12 @@ static void move_expired_inodes(struct l
 		if (older_than_this &&
 		    inode_dirtied_after(inode, *older_than_this))
 			break;
+		list_move(&inode->i_wb_list, &tmp);
+		if (inode->i_sb == blockdev_superblock)
+			continue;
 		if (sb && sb != inode->i_sb)
 			do_sb_sort = 1;
 		sb = inode->i_sb;
-		list_move(&inode->i_wb_list, &tmp);
 	}
 
 	/* just one sb in list, splice to dispatch_queue and we're done */
